csr_check.go

```diff
@@ -162,3 +197,110 @@ func authorizeCSR(machineList *v1beta1.MachineList, req *certificatesv1beta1.Cer
 
 	return nil
 }
+
+func authorizeNodeClientCSR(machines []v1beta1.Machine, nodes corev1client.NodeInterface, req *certificatesv1beta1.CertificateSigningRequest, csr *x509.CertificateRequest) error {
+	if !isReqFromNodeBootstrapper(req) {
+		return fmt.Errorf("CSR %s for node client cert has wrong user", req.Name)
+	}
+
+	nodeName := strings.TrimPrefix(csr.Subject.CommonName, nodeUserPrefix)
+	if len(nodeName) == 0 {
+		return fmt.Errorf("CSR %s has empty node name", req.Name)
+	}
+
+	_, err := nodes.Get(nodeName, metav1.GetOptions{})
+	switch {
+	case err == nil:
+		return fmt.Errorf("node %s already exists", nodeName)
+	case errors.IsNotFound(err):
+		// good, node does not exist
+	default:
+		return fmt.Errorf("failed to check if node %s already exists: %v", nodeName, err)
+	}
+
+	nodeMachine, ok := findMatchingMachineFromInternalDNS(nodeName, machines)
+	if !ok {
+		return fmt.Errorf("failed to find machine for node %s", nodeName)
+	}
+
+	if nodeMachine.Status.NodeRef != nil {
+		return fmt.Errorf("machine for node %s already has node ref", nodeName)
+	}
+
+	if !machineWantsCSRAutoApprove(nodeMachine) {
+		return fmt.Errorf("machine for node %s is not configured for CSR auto approval", nodeName)
+	}
+
+	start := nodeMachine.CreationTimestamp.Add(-maxMachineClockSkew)
+	end := nodeMachine.CreationTimestamp.Add(maxMachineDelta)
+	if !inTimeSpan(start, end, req.CreationTimestamp.Time) {
+		return fmt.Errorf("CSR %s creation time %s not in range (%s, %s)", req.Name, req.CreationTimestamp.Time, start, end)
+	}
+
+	return nil // approve node client cert
+}
+
+func isReqFromNodeBootstrapper(req *certificatesv1beta1.CertificateSigningRequest) bool {
+	return req.Spec.Username == nodeBootstrapperUsername && nodeBootstrapperGroups.Equal(sets.NewString(req.Spec.Groups...))
+}
+
+func findMatchingMachineFromNodeRef(nodeName string, machines []v1beta1.Machine) (v1beta1.Machine, bool) {
+	for _, machine := range machines {
+		if machine.Status.NodeRef != nil && machine.Status.NodeRef.Name == nodeName {
+			return machine, true
+		}
+	}
+	return v1beta1.Machine{}, false
+}
+
+func findMatchingMachineFromInternalDNS(nodeName string, machines []v1beta1.Machine) (v1beta1.Machine, bool) {
+	for _, machine := range machines {
+		for _, address := range machine.Status.Addresses {
+			if address.Type == corev1.NodeInternalDNS && address.Address == nodeName {
+				return machine, true
+			}
+		}
+	}
+	return v1beta1.Machine{}, false
+}
+
+func machineWantsCSRAutoApprove(machine v1beta1.Machine) bool {
+	return machine.Labels[machineCSRAutoApproveKey] == machineCSRAutoApproveValue
```

Comment:
just want to know pros/cons of label versus annotation.

i tend to think of labels as grouping, and annotations as behavioral.